File: 0200-number-of-islands/0200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        def dfs(r,c):
            if min(r,c) < 0 or r == ROWS or c == COLS or grid[r][c] == "2" or grid[r][c] == "0":
                return
            
            grid[r][c] = "2"
            
            dfs(r-1, c)
            dfs(r, c-1)
            dfs(r+1,c)
            dfs(r,c+1)
        
        
        ROWS, COLS = len(grid), len(grid[0])
        res = 0
        
        for r in range(ROWS):
            for c in range(COLS):
                if grid[r][c] == "1":
                    res += 1
                    dfs(r,c)
                c += 1
            r+= 1
        
        return res
        
        
        
            
            
            
            
            
        

File: 0200-number-of-islands/0200-number-of-islands.cpp
class Solution {
public:

    void dfs(vector<vector<char>>& grid, int r, int c) {
        int rows = grid.size();
        int cols = grid[0].size();

        if ((r < 0) || (c < 0) || (r >= rows) || (c >= cols) || grid[r][c] == '0') return;
        else {
            grid[r][c] = '0';
            dfs(grid, r + 1, c);
            dfs(grid, r - 1, c);
            dfs(grid, r, c + 1);
            dfs(grid, r, c - 1);
        }
        return;
    }

    int numIslands(vector<vector<char>>& grid) {
        int rows = grid.size();
        int cols = grid[0].size();
        int counter =0;
        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                if (grid[r][c] == '1'){
                    dfs(grid, r, c);
                    counter++;
                }
            }
        }
        return counter;
    }
};